Update ontoportal docker to use docker compose of the ui and api

imadbourouche opened the pull request:
## Prequisities 
* https://github.com/ontoportal-lirmm/ontologies_api/pull/59
* https://github.com/ontoportal-lirmm/bioportal_web_ui/pull/386

## Proposition 
The Ontoportal_docker repository has simplified the way ontoportal development works by containerizing the appliance component (API, UI ...etc). 

The main goal of this PR is to make a one-click solution designed to simplify the Running and deployment of the Ontoportal appliance. This script streamlines the process of setting up and running the Ontologies API and User Interface (UI) components on your local machine or deploying it on a remote server.

This script also gives the user the ability to make his own configuration that has to be in the .env file. So it has two modes:
- default mode: where the script will pull the configuration from `ontoportal-lirmm`
- personalized mode: where the script will use the `.env` file provided by the user


You can find here the full documentation: [Documentation](https://www.notion.so/imadbourouche/Ontortal-Docker-Documentation-79f6d0c8662d48da87780fca6f2e36da)

Demonstration Video: [Demo](https://drive.google.com/drive/folders/13KviBGEKJK4pp1u1MMgO4rSGnE8k_Eoh?usp=sharing)

**How it works**

just run `./ontoportal help` to see the help
```
➜  ontoportal_docker git:(main) ./ontoportal help
Usage: ./ontoportal <command>

Commands:
  start   : Start the ontologies API and UI (Default Mode / Personilized Mode)
  deploy  : Deploy the application using Kamal to the server
  stop    : Stop the API and UI services
  clean   : Clean up the server (remove ontologies_linked_data, docker-compose files, and remove all containers)
  help    : Show this help message

Additional notes:
- You can provide your own .env file to customize parameters:
  Just put the .env file in this directory and run ./ontoportal start
```

**Run The appliance (API + UI)**
```
./ontoportal start
```

**Stop The appliance (API + UI)**
```
./ontoportal stop
```

**Clean The server**: this will remove all docker-compose files and containers
```
./ontoportal clean
```

**Deploy The appliance on a server**: this will deploy the api and the ui on a server using configuration in .env file
```
./ontoportal deploy
```

imadbourouche committed:
Update ontoportal docker to use docker compose of the ui and api

using ontoportal script

imadbourouche committed:
Make the script more configurable using .env file and simplify the run_api and run_ui scripts

syphax-bouazzouni commented on ontoportal:
```diff
@@ -0,0 +1,238 @@
+	      echo "[+] Directory $repo_dir already exists. Skipping cloning."
+	  else
+	    echo "[+] Cloning ontologies_linked_data repo..."
+	    git clone -q --depth=1 https://github.com/ontoportal-lirmm/ontologies_linked_data.git
```
this need to be configurable

syphax-bouazzouni commented on ontoportal:
```diff
@@ -0,0 +1,238 @@
+		echo "[-] Error: Missing required arguments. Please provide both --api-url or update them in your .env"
+		exit 1
+	fi
+	bash_cmd=" (bundle check || bundle install || bundle update) && bundle exec rackup --host 0.0.0.0  --env=development --port 9393"
```
if you remove `(bundle check || bundle install || bundle update)`  does it work ?

syphax-bouazzouni commented on ontoportal:
```diff
@@ -0,0 +1,238 @@
+		exit 1
+	fi
+	echo "[+] Starting the UI"
+	bash_cmd="cp /app/config/bioportal_config_env.rb.sample /app/config/bioportal_config_development.rb && cp /app/config/database.yml.sample /app/config/database.yml && (bundle check || bundle install) && bin/rails db:prepare && bundle exec rails s -b 0.0.0.0 -p 3000"
```
if you remove `(bundle check || bundle install || bundle update)` does it work

syphax-bouazzouni commented on ontoportal:
```diff
@@ -0,0 +1,238 @@
+
+	run_api_default_mode(){
+		echo "[+] Getting compose file for API"
+		curl  -sS -L https://raw.githubusercontent.com/ontoportal-lirmm/ontologies_api/development/docker-compose.yml -o docker-compose_api.yml
```
this need to be configurable

syphax-bouazzouni commented on ontoportal:
```diff
@@ -0,0 +1,238 @@
+		echo "[+] Getting compose file for API"
+		curl  -sS -L https://raw.githubusercontent.com/ontoportal-lirmm/ontologies_api/development/docker-compose.yml -o docker-compose_api.yml
+		echo >> .env
+		curl  -sS -L https://raw.githubusercontent.com/ontoportal-lirmm/ontologies_api/development/.env.sample >> .env
```
this need to be configurable

syphax-bouazzouni commented on ontoportal:
```diff
@@ -0,0 +1,238 @@
+
+	run_ui_default_mode(){
+		echo "[+] Getting compose file for UI"
+		curl -sS -L https://raw.githubusercontent.com/ontoportal-lirmm/bioportal_web_ui/development/docker-compose.yml -o docker-compose_ui.yml
```
this need to be configurable

syphax-bouazzouni commented on ontoportal:
```diff
@@ -0,0 +1,238 @@
+		echo "[+] Getting compose file for UI"
+		curl -sS -L https://raw.githubusercontent.com/ontoportal-lirmm/bioportal_web_ui/development/docker-compose.yml -o docker-compose_ui.yml
+		echo >> .env
+		curl -sS -L https://raw.githubusercontent.com/ontoportal-lirmm/bioportal_web_ui/development/.env.sample >> .env
```
this need to be configurable

syphax-bouazzouni commented on ontoportal:
```diff
@@ -0,0 +1,238 @@
+		curl -sS -L https://raw.githubusercontent.com/ontoportal-lirmm/bioportal_web_ui/development/.env.sample >> .env
+		
+		sed -i 's/^API_URL=http:\/\/localhost:9393$/API_URL=http:\/\/api-service:9393/' .env
+		sed -i 's/^API_KEY=$/API_KEY=72c72cba-ad45-4785-b94e-483fa55cdddb/' .env
```
the APIKEY need to be configurable

syphax-bouazzouni commented on ontoportal:
```diff
@@ -0,0 +1,238 @@
+  echo "[+] Removing docker compose files"
+  rm -f docker-compose*
+  echo "[+] Removing All containers/volumes/networks"
+  docker container rm -f $(docker ps -aq) 2>/dev/null; docker volume rm $(docker volume ls -q) 2>/dev/null ; docker network prune -f
```
this clean needs to affect only images/containers related to the current running ontoportal, not removing everything.
As in a server, you may have multiple applications using docker.
Also, the volumes of the database, need to be persistent, we don't want to remove them on each deployment or start.

syphax-bouazzouni commented:
Also, you(@imadbourouche) need to add the `.env.sample` file of the ontoportal docker

imadbourouche committed:
Configure ontologies_linked_data path & remove (bundle check || bundle install || bundle update) from run_api

imadbourouche committed:
Fix the clear methode to clean only the ontoportal_docker running and to not remove the volumes

imadbourouche committed:
Get the env file by copying .env.sample not by pulling from the repositories

imadbourouche committed:
Enhanced script configurability and personalization

- Added support for user-defined environment variables
- Give the user ability to:
   - start/deploy the services that he wants (api / ui / ontoportal)
   - specify the source of his project (docker compose files)
- Fixing some functionalities:
   - fix the clean/stop methods to clean/stop only the services related to our project
- Refactored code for better readability and maintainability
